lib/message-handler.js

```diff
@@ -906,452 +915,483 @@ class MessageHandler {
+                            uid: messageUid
+                        },
+                        { writeConcern: 'majority' }
+                    );
+                } catch (err) {
+                    await cursor.close();
+                    return done(err);
+                }
+
+                if (deleteMessageRes && deleteMessageRes.deletedCount) {
+                    if (options.session) {
+                        options.session.writeStream.write(options.session.formatResponse('EXPUNGE', sourceUid));
+                    }
+
+                    removeEntries.push({
+                        command: 'EXPUNGE',
+                        ignore: options.session && options.session.id,
+                        uid: messageUid,
+                        message: messageId,
+                        thread: message.thread,
+                        unseen,
+                        // modseq is needed to avoid updating mailbox entry
+                        modseq: newModseq
+                    });
+
+                    if (options.showExpunged) {
+                        options.session.writeStream.write(options.session.formatResponse('EXPUNGE', messageUid));
+                    }
+                }
+
+                let entry = {
+                    command: 'EXISTS',
+                    uid: uidNext,
+                    message: insertId,
+                    unseen: message.unseen,
+                    idate: message.idate,
+                    thread: message.thread
+                };
+                if (junk) {
+                    entry.junk = junk;
+                }
+                existsEntries.push(entry);
+
+                if (existsEntries.length >= consts.BULK_BATCH_SIZE) {
+                    // mark messages as deleted from old mailbox
+                    return new Promise(resolve => {
+                        this.notifier.addEntries(mailboxData, removeEntries, () => {
+                            // mark messages as added to new mailbox
+                            this.notifier.addEntries(targetData, existsEntries, () => {
+                                removeEntries = [];
+                                existsEntries = [];
+                                this.notifier.fire(mailboxData.user);
+                                resolve(processNext());
+                            });
+                        });
+                    });
+                }
+                return processNext();
+            };
+
+            if (targetData.encryptMessages) {
+                // move target mailbox is encrypted
+                const parsedHeader = (message.mimeTree && message.mimeTree.parsedHeader) || {};
+                const parsedContentType = parsedHeader['content-type'];
+
+                if (parsedContentType && parsedContentType.subtype === 'encrypted') {
+                    // message already encrypted, just continue move
+                    return updateMessage();
```

Comment:
`updateMessage` is an async function so it is hard to say what exactly is returned here and other places (a Promise). Maybe always await it before returning?